### PR TITLE
All Remaining Types for a First Working Version

### PR DESCRIPTION
**Pull request by vvolkl**
BEGINRELEASENOTES
- Add plcio-like types and some associations needed for a first working version: TrackState, ObjectID, TrackerHit, TPCHit, Track, Vertex, ReconstructedParticle, MCRecoParticleAssociation, MCRecoTrackerAssociation, MCRecoCaloAssociation

ENDRELEASENOTES

### CODE REVIEW

File: edm4hep.yaml [hunk 1]
```diff
@@ -46,6 +46,38 @@ components:
         bool operator==(const Vector2i& v) const { return (a==v.a&&b==v.b) ; }\n
         int operator[](unsigned i) const { return *( &a + i ) ; }\n
         "
+      # Parametrized description of a particle track
+    edm4hep::TrackState:
+      location       : int # for use with At{Other|IP|FirstHit|LastHit|Calorimeter|Vertex|LastLocation}
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      location       : int # for use with At{Other|IP|FirstHit|LastHit|Calorimeter|Vertex}|LastLocation
```
LastLocation does not start with `At`

File: edm4hep.yaml [hunk 2]
```diff
@@ -135,11 +167,17 @@ datatypes :
...
+      "
       const_declaration: "
       static const int  BITOverlay = 31;\n
       static const int  BITProducedBySecondary = 30;\n
       bool isOverlay() const { return getQuality() & (1 << BITOverlay) ; }\n
       bool isProducedBySecondary() const { return getQuality() & (1 << BITProducedBySecondary) ; }\n
+      double x() const {return getPosition()[0];};\n
+      double y() const {return getPosition()[1];};\n
+      double z() const {return getPosition()[2];};\n
+      double rho() const {return sqrt(pow(getPosition()[1],2) + pow(getPosition()[0],2));};\n
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      double rho() const {return sqrt(x()*x() + y()*y());}\n
```

File: edm4hep.yaml [hunk 3]
```diff
@@ -135,11 +167,17 @@ datatypes :
...
       - edm4hep::MCParticle MCParticle //MCParticle that caused the hit.
     ExtraCode :
+      includes: "#include <math.h>\n
+      "
       const_declaration: "
       static const int  BITOverlay = 31;\n
       static const int  BITProducedBySecondary = 30;\n
       bool isOverlay() const { return getQuality() & (1 << BITOverlay) ; }\n
       bool isProducedBySecondary() const { return getQuality() & (1 << BITProducedBySecondary) ; }\n
+      double x() const {return getPosition()[0];};\n
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      double x() const {return getPosition()[0];}\n
```

File: edm4hep.yaml [hunk 4]
```diff
@@ -135,11 +167,17 @@ datatypes :
...
+      double y() const {return getPosition()[1];};\n
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      double y() const {return getPosition()[1];}\n
```

File: edm4hep.yaml [hunk 5]
```diff
@@ -135,11 +167,17 @@ datatypes :
...
+      double z() const {return getPosition()[2];};\n
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      double z() const {return getPosition()[2];}\n
```

File: edm4hep.yaml [hunk 6]
```diff
@@ -227,3 +265,118 @@ datatypes :
...
+    Description : "Tracker hit"
+    Author : "F.Gaede, DESY"
+    Members :
+      - unsigned long long cellID      //ID of the sensor that created this hit
+      - int type                       //type of raw data hit, either one of edm4hep::TPCHIT, edm4hep::SIMTRACKERHIT - see collection parameters "TrackerHitTypeNames" and "TrackerHitTypeValues". 
+      - int quality                    //quality bit flag of the hit.    
+      - float time                     //time of the hit.
+      - float eDep                     //energy deposited on the hit [GeV]. 
+      - float eDepError                //error measured on EDep [GeV].
+      - float edx                      //dE/dx of the hit in [GeV].
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      - float edx                      //dE/dx of the hit in [GeV/mm]. 
```
?

File: edm4hep.yaml [hunk 7]
```diff
@@ -227,3 +265,118 @@ datatypes :
...
+      - int ndf                          //number of degrees of freedom of the track fit
+      - float dEdx                       //dEdx of the track.
+      - float dEdxError                  //error of dEdx.
+      - float radiusOfInnermostHit       //radius of the innermost hit that has been used in the track fit
+    VectorMembers:
+      - int subDetectorHitNumbers        //number of hits in particular subdetectors.Check/set collection variable TrackSubdetectorNames for decoding the indices 
+      - edm4hep::TrackState trackStates  //track states
+    OneToManyRelations:
+      - edm4hep::TrackerHit trackerHits  //hits that have been used to create this track
+      - edm4hep::Track tracks            //tracks (segements) that have been combined to create this track
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      - edm4hep::Track tracks            //tracks (segments) that have been combined to create this track
```

File: edm4hep.yaml [hunk 8]
```diff
@@ -227,3 +265,118 @@ datatypes :
...
+
+ #---------- Vertex
+  edm4hep::Vertex:
+    Description: "LCIO vertex"
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
    Description: "Vertex"
```

File: edm4hep.yaml [hunk 9]
```diff
@@ -227,3 +265,118 @@ datatypes :
...
+      - edm4hep::Vertex          startVertex    //start vertex associated to this particle
+      - edm4hep::ParticleID      particleIDUsed //particle Id used for the kinematics of this particle
+    OneToManyRelations:
+      - edm4hep::Cluster               clusters     //clusters that have been used for this particle. 
+      - edm4hep::Track                 tracks       //tracks that have been used for this particle. 
+      - edm4hep::ReconstructedParticle particles    //reconstructed particles that have been combined to this particle.
+      - edm4hep::ParticleID            particleIDs  //particle Ids (not sorted by their likelihood) 
+    ExtraCode:
+      declaration: "
+      bool isCompound() { return particles_size() > 0 ;}\n
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
      bool isCompound() { return particles.size() > 0 ;}\n
```
? Or is this some special podio construct?

File: edm4hep.yaml [hunk 10]
```diff
@@ -227,3 +265,118 @@ datatypes :
...
+
+  edm4hep::MCRecoCaloAssociation:
+    Description: "Association between a CaloHit and the corresponding simulated CaloHit"
+    Author: "C. Bernet, B. Hegner"
+    OneToOneRelations:
+     - edm4hep::CalorimeterHit rec // The reconstructed hit.
+     - edm4hep::SimCalorimeterHit sim // The simulated hit.
+
+  edm4hep::MCRecoTrackerAssociation:
+    Description: "Association between a TrackerHit and the corresponding simulated CaloHit"
```
Comment (andresailer) [inline edm4hep.yaml]:
```suggestion
    Description: "Association between a TrackerHit and the corresponding simulated TrackerHit"
```

File: edm4hep.yaml (hunk 9 above)
Comment (vvolkl) [reply edm4hep.yaml]: Indeed, this is ok I think. Podio will generate a `${name}_size()` for these types of relations. I believe it comes from [here](https://github.com/AIDASoft/podio/blob/ad31c36f077487a1da130d3297ddcbb74f19e62a/python/templates/ConstRefVector.h.template). We should maybe also put the [generated code](https://gist.github.com/vvolkl/fc5d912ce6ba62840e512b4dfc89a4eb) for these classes on a doxygen page somewhere.

File: edm4hep.yaml (hunk 6 above)
Comment (vvolkl) [reply edm4hep.yaml]: @gaede I believe Andre is correct, right? The comments in LCIO are sometimes [GeV](https://github.com/iLCSoft/LCIO/blob/568e2075843ea8c6ef3ee852a591633dedc85a17/src/cpp/src/UTIL/Operators.cc#L1780) and sometimes [GeV/mm](https://github.com/iLCSoft/LCIO/blob/27c67df622c934f3258a38e167b4b358781f306a/src/cpp/include/IMPL/TrackerHitZCylinderImpl.h#L66)

**Comment by vvolkl** [comment]
Thank you @andresailer for the comments!

**Comment by vvolkl** [comment]
Thanks Gerri, agreed, I created #25 to keep track of this issue, so we can go ahead and merge.

**Comment by vvolkl** [comment]
I'm still working to fix the CI, but the [build on my fork](https://github.com/vvolkl/EDM4HEP/runs/425645508) passed.

**Pull request merged**